File: app/src/main/java/dev/ragnarok/fenrir/view/emoji/SectionsAdapter.java
package dev.ragnarok.fenrir.view.emoji;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.picasso.PicassoInstance;
import dev.ragnarok.fenrir.settings.CurrentTheme;

public class SectionsAdapter extends RecyclerView.Adapter<SectionsAdapter.Holder> {

    private final Context mContext;
    private final List<AbsSection> data;
    private Listener listener;

    public SectionsAdapter(List<AbsSection> data, Context mContext) {
        this.data = data;
        this.mContext = mContext;
    }

    @NonNull
    @Override
    public Holder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new Holder(LayoutInflater.from(mContext).inflate(R.layout.emoji_section_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull Holder holder, int position) {
        AbsSection section = data.get(position);
        switch (section.type) {
            case AbsSection.TYPE_EMOJI:
                EmojiSection emojiSection = (EmojiSection) section;

                PicassoInstance.with()
                        .cancelRequest(holder.icon);

                holder.icon.setImageDrawable(emojiSection.drawable);
                holder.icon.getDrawable().setTint(CurrentTheme.getColorOnSurface(mContext));
                break;

            case AbsSection.TYPE_STICKER:
                StickerSection stickerSection = (StickerSection) section;
                if (stickerSection.stickerSet.getTitle() != null && stickerSection.stickerSet.getTitle().equals("recent")) {
                    holder.icon.setImageResource(R.drawable.pin);
                    holder.icon.getDrawable().setTint(CurrentTheme.getColorPrimary(mContext));
                } else {
                    PicassoInstance.with()
                            .load(stickerSection.stickerSet.getImageUrl(128))
                            .placeholder(R.drawable.sticker_pack_with_alpha)
                            .into(holder.icon);
                    holder.icon.setColorFilter(null);
                }
                break;
            case AbsSection.TYPE_PHOTO_ALBUM:
                PicassoInstance.with()
                        .cancelRequest(holder.icon);

                holder.icon.setImageResource(R.drawable.image);
                holder.icon.getDrawable().setTint(CurrentTheme.getColorOnSurface(mContext));
                break;
        }

        if (section.active) {
            holder.root.setBackgroundResource(R.drawable.circle_back_white);
            holder.root.getBackground().setTint(CurrentTheme.getMessageBackgroundSquare(mContext));
        } else {
            holder.root.setBackground(null);
        }

        holder.itemView.setOnClickListener(v -> {
            if (listener != null) {
                listener.onClick(holder.getBindingAdapterPosition());
            }
        });
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public void setListener(Listener listener) {
        this.listener = listener;
    }

    public interface Listener {
        void onClick(int position);
    }

    public static class Holder extends RecyclerView.ViewHolder {

        final View root;
        final ImageView icon;

        public Holder(View itemView) {
            super(itemView);
            root = itemView.findViewById(R.id.root);
            icon = itemView.findViewById(R.id.icon);
        }
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/view/emoji/SectionsAdapter.kt
package dev.ragnarok.fenrir.view.emoji

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.recyclerview.widget.RecyclerView
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.picasso.PicassoInstance.Companion.with
import dev.ragnarok.fenrir.settings.CurrentTheme

class SectionsAdapter(private val data: List<AbsSection>, private val mContext: Context) :
    RecyclerView.Adapter<SectionsAdapter.Holder>() {
    private var listener: Listener? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder {
        return Holder(
            LayoutInflater.from(mContext).inflate(R.layout.emoji_section_item, parent, false)
        )
    }

    override fun onBindViewHolder(holder: Holder, position: Int) {
        val section = data[position]
        when (section.type) {
            AbsSection.TYPE_EMOJI -> {
                val emojiSection = section as EmojiSection
                with()
                    .cancelRequest(holder.icon)
                holder.icon.setImageDrawable(emojiSection.drawable)
                holder.icon.drawable.setTint(CurrentTheme.getColorOnSurface(mContext))
            }
            AbsSection.TYPE_STICKER -> {
                val stickerSection = section as StickerSection
                if (stickerSection.stickerSet.getTitle() != null && stickerSection.stickerSet.getTitle() == "recent") {
                    holder.icon.setImageResource(R.drawable.pin)
                    holder.icon.drawable.setTint(CurrentTheme.getColorPrimary(mContext))
                } else {
                    with()
                        .load(stickerSection.stickerSet.getImageUrl(128))
                        .placeholder(R.drawable.sticker_pack_with_alpha)
                        .into(holder.icon)
                    holder.icon.colorFilter = null
                }
            }
            AbsSection.TYPE_PHOTO_ALBUM -> {
                with()
                    .cancelRequest(holder.icon)
                holder.icon.setImageResource(R.drawable.image)
                holder.icon.drawable.setTint(CurrentTheme.getColorOnSurface(mContext))
            }
        }
        if (section.active) {
            holder.root.setBackgroundResource(R.drawable.circle_back_white)
            holder.root.background.setTint(CurrentTheme.getMessageBackgroundSquare(mContext))
        } else {
            holder.root.background = null
        }
        holder.itemView.setOnClickListener {
            listener?.onClick(holder.bindingAdapterPosition)
        }
    }

    override fun getItemCount(): Int {
        return data.size
    }

    fun setListener(listener: Listener?) {
        this.listener = listener
    }

    interface Listener {
        fun onClick(position: Int)
    }

    class Holder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val root: View = itemView.findViewById(R.id.root)
        val icon: ImageView = itemView.findViewById(R.id.icon)

    }
}